### ISSUE TITLE
Implement a Turtle and TriG parser

### ISSUE THREAD

**Issue by Tpt** [Tpt/rudf#3]
Specs:
* https://www.w3.org/TR/turtle/
* https://www.w3.org/TR/trig/

Testsuite: https://www.w3.org/TR/rdf11-testcases/

**Comment by tobias47n9e** [comment]
Were you thinking of a serde extension like serdejson for this?

**Comment by Tpt** [comment]
serde is good for simple (de)serialization of arbitrary datastructure. Here we want to parse a comple language in a small set of weel-known datastructure. I believe that something like [rust-peg](https://github.com/kevinmehall/rust-peg) that we already use for N-Triple parsing would be better, at least for a first implementation.

**Comment by XAMPPRocky** [comment]
As I'm sure you're aware (since I found this repo through chelone being starred) I already have written a [Turtle parser in Rust](https://github.com/XAMPPRocky/Chelone). I haven't worked on it recently due to other time commitments and I didn't want to write the rest of the RDF stack in order to have a client to improve its API. I am still willing to maintain it however. It's pretty complete, of the 291 compliance tests, it fails about 7 of them currently. I would like to continue the work to make it fully compliant. I am also open to any API suggestions & improvements as the current API was designed around these compliance tests, and not real world clients.

**Comment by Tpt** [comment]
@XAMPPRocky Hi! Thank you very much for your message. I have actually started to write a new Turtle parser in plain Rust on my own (Rudf already contains one but it was very slow). I just finished [a 0.1 version that passes all the compliance tests](https://crates.io/crates/rio_turtle) and migrated Rudf to use it.

The source code is [here](https://github.com/Tpt/rio) and some benches with a comparison with Chelone are [here in the `bench` branch](https://github.com/Tpt/rio/blob/bench/testsuite/benches/w3c_testsuite.rs).

My parser seems already to be fairly fast according to this benchmark: it parses the concatenation of all valid Turtle files from the compliance tests at 82 MB/s. I have not done any particular optimization at this step, so it's hopefully possible to speed it up more.
